### DISCUSSION TITLE
Burn and .NET Framework 4.8

### DISCUSSION THREAD

**Discussion by TrulsHaraldsson** [wixtoolset/issues#6713] [Questions] [answered]
Hi Wix Community!

**TLDR: Two web installers, 1 working, 1 not working. Cannot understand the problem.**

This might turn out to be a "non-wix-question", but anyways. We have an installer that requires the users machine to have .NET Framework 4.8 installed. So I created a Wix Bundle and included a web installer for the .NET Framework as 
```
     <!--
        This package is for .NET 4.8 and will be pulled down from the internet if it not installed. This
        mainly concerns users on windows 7 machines.
      -->
      <ExePackage
        SourceFile="Resources\ndp48-web.exe"
        Permanent="yes"
        PerMachine="yes"
        InstallCommand="/norestart /passive /showrmui"
      />
```
this worked well, at least for a while. Before I continue I would like to say, that I wrote this code more than half a year ago, at which point we tested the installer on the target systems (Windows 7, 8, 10 etc).

One week ago we got a customer report that their installer was failing with the reported error:
![image](https://user-images.githubusercontent.com/4319850/152348933-1d85eeb7-05a7-462f-af25-eb3fbac1d5aa.png)

So I booted up the virtual machine for Windows 7 (same snapshot) as I used when I first verified the installer, and to my surprise I got the same error. Looking in the log I found the following rows
```
[0EB4:0EB8][2022-02-03T02:42:54]i301: Applying execute package: ndp48_web.exe, action: Install, path: C:\ProgramData\Package Cache\5A84A8E612E270E27D0061D58DB6B470153BE1F9\ndp48-web.exe, arguments: '"C:\ProgramData\Package Cache\5A84A8E612E270E27D0061D58DB6B470153BE1F9\ndp48-web.exe" /norestart /passive /showrmui'
[0EB4:0EB8][2022-02-03T02:46:03]e000: Error 0x800b0101: Process returned error: 0x800b0101
[0EB4:0EB8][2022-02-03T02:46:03]e000: Error 0x800b0101: Failed to execute EXE package.
[0E7C:0E80][2022-02-03T02:46:03]e000: Error 0x800b0101: Failed to configure per-machine EXE package.
[0E7C:0E80][2022-02-03T02:46:03]i319: Applied execute package: ndp48_web.exe, result: 0x800b0101, restart: None
[0E7C:0E80][2022-02-03T02:46:03]e000: Error 0x800b0101: Failed to execute EXE package.
```

So I extracted the .NET installer (i.e. npd48-web.exe) from the bundle to run it alone to see if I could get more information. In the log for npd48-web.exe installer I find this
```
Downloading http://go.microsoft.com/fwlink/ to C:\1e6ec246466378d50f18af8225e22a\netfx_Full_x86.msi
Action: Downloading Item 
Starting download attempt 1 of 4 for http://go.microsoft.com/fwlink/?prd=11324&pver=netfx&sbp=Net48Rel1&plcid=0x409&clcid=0x409&ar=03761.00&sar=amd64&o1=netfx_Full_x86.msi using BITS
 Log File C:\Users\IEUser\AppData\Local\Temp\dd_BITS.log does not yet exist but may do at Watson upload time
 Log File C:\Users\IEUser\AppData\Local\Temp\dd_BITS.log does not yet exist but may do at Watson upload time
Action: Downloading http://go.microsoft.com/fwlink/?prd=11324&pver=netfx&sbp=Net48Rel1&plcid=0x409&clcid=0x409&ar=03761.00&sar=amd64&o1=netfx_Full_x86.msi using BITS 
Action complete
C:\1e6ec246466378d50f18af8225e22a\TMP7F1C.tmp: Verifying signature for netfx_Full_x86.msi
Signature verified, verifying signer
 Failed to find countersignature OID
 GetDateFromCryptMsg failed with error: 2148204811
 CertVerifyCertificateChainPolicy failed with error: 2148204801
 C:\1e6ec246466378d50f18af8225e22a\TMP7F1C.tmp - Signature verification for file netfx_Full_x86.msi (C:\1e6ec246466378d50f18af8225e22a\TMP7F1C.tmp) failed with error 0x800b0101 (A required certificate is not within its validity period when verifying against the current system clock or the timestamp in the signed file.)
Crypto API Events: 
EvtNext failed with error: 4317
 <events/>
Crypto API Events 
C:\1e6ec246466378d50f18af8225e22a\TMP7F1C.tmp Signature could not be verified for netfx_Full_x86.msi
 No FileHash provided. Cannot perform FileHash verification for netfx_Full_x86.msi
 Download failed at attempt 1 of 4 for http://go.microsoft.com/fwlink/?prd=11324&pver=netfx&sbp=Net48Rel1&plcid=0x409&clcid=0x409&ar=03761.00&sar=amd64&o1=netfx_Full_x86.msi using BITS
```
In the log it tries 3 more times before the installer fails. Each time it fails at verify the signature, and we can see the same reported message
`A required certificate is not within its validity period when verifying against the current system clock or the timestamp in the signed file.`

So I try the obvious, setting the system clock. But at the time of writing this post the clock seemed to match my timezone and time 
![image](https://user-images.githubusercontent.com/4319850/152355071-5554cef3-7ee8-4fb6-853e-d848510919c2.png)

To further investigate the problem I decided to download a new web installer from Microsoft's [website](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net48) (I downloaded the one we use today from the same place) and tried running that one instead. To my surprise that installer worked.

So here I am with two version of the net 48 web installer, the one we have (not working) 
![image](https://user-images.githubusercontent.com/4319850/152356614-d9660923-1302-4b28-a129-c9a33f9b6a61.png) 
and the newly downloaded (working)
![image](https://user-images.githubusercontent.com/4319850/152357445-f5d371c5-7fcf-4a35-b25c-c21c33d4d226.png)

The workaround right now is to switch the non-working to the working installer, but I'm afraid that the now working installer will stop working in the future as well.

Do you guys have any idea what the problem could be or maybe what the next step could be? 

I'm still new to Wix (~1 year experience) and I might have overlooked something here. Like should I use a web installer like I do?

Thanks in advance.

**Comment by robmen** [comment, marked as answer]
Very likely this issue: A little while back, Microsoft decided to release all the .NET Framework binaries with new signatures at the same URLs. This broke everyone validating the .NET Framework thumbprint, including existing Burn bundles. It was pretty disruptive. We had to help a couple of FireGiant customers navigate emergency updates.

Could it happen again? Yes. Microsoft could choose to make a breaking change and post it to the same URL. Will they? They didn't do it for many years, never was clear why they did it this time.

**Reply by TrulsHaraldsson** [reply]
Thanks for the quick response Rob! If that is the case I feel a lot more confident with replacing our current web installer with the new one.
